Minimalist example with scikit-learn (#339)
Epistimio/orion: docs/src/examples/scikit-learn.rst

********************
Scikit-learn
********************
.. Might also think of moving this file to examples/ then we an example auto contained in the
   repository. We invoke this file from index.rst

In this example, we're going to demonstrate how Oríon can be integrated to a minimal model using

> scikit-learn [scikit]_ on the iris dataset [iris]_. The files mentioned in this example are

Reviewer: I think it would be more useful here to provide a URL to scikit instead of a citation. For iris it would be great to provide an url to the dataset as well. Similarly, the line below should point to the examples folder in the repo on github, so the user can directly go see.

Author: Good catch! I'm going to change it.

available in the folder :file:`examples/scikitlearn-iris/`.

The requirements are listed in requirements.txt. You can quickly install them using ``$ pip
install -r requirements.txt``. If you haven't installed Oríon previously, make sure to
:doc:`configure it properly <../install/core>` before going further.

Sample script
---------------

.. literalinclude:: ../../../examples/scikitlearn-iris/main.py
   :language: python
   :lines: 1-2, 5-9, 13-30

This very basic script takes in parameter one positional argument for the hyper-parameter *epsilon*
which control the loss in the script.

The script is divided in five parts:

#. Parsing of the script arguments
#. Loading and splitting the dataset
#. Training a classifier using the researcher-defined *epsilon*
#. Evaluating the classifier using the testing set
#. Reporting the performance of the model. i.e., the accuracy.

.. note::

>    The workflow presented in the script is simplified compared to the reality on purpose. The

Reviewer: ```suggestion
   The workflow presented in the script is simplified on purpose compared to real ones. The
```

   objective of this example is to illustrate the basic steps involved in using Oríon.

To find a good *epsilon*, a user would run empirically ``$ python main.py <epsilon>`` multiple
times, choosing a new value for *epsilon* manually.

This ad-hoc hyper-parameter optimization is unreliable, slow, and requires a lot of work from the

> user. Oríon solves this problem by providing cutting-edge hyper-parameter optimization

Reviewer: ```suggestion
user. Oríon solves this problem by providing established hyper-parameter optimization
```
We prioritize established and reliable algorithms.

Author: Good catch! Thanks!

algorithms without disrupting the workflow of the user. Integrating it only require minimal
adjustments to your current workflow as we'll demonstrate in the next section.

Enter Orion
-----------
Integrating Oríon into your workflow requires only two non-invasive changes:
   1. Define an objective to optimize.
   2. Specify the hyper-parameter space.

For the former, this step takes place in the script training the model. The latter can either be
specified in a configuration file or directly while calling the script with Oríon.
For the purpose of the example, we'll configure the hyper-parameter space directly as a
command-line argument.

Updating the script
^^^^^^^^^^^^^^^^^^^
We only need to make one small change to the script: we report to Oríon the objective that we

> want to **minimize** at the end of the script using :py:func:`orion.client.report_results`:

Reviewer: ```suggestion
want to **minimize** at the end of the script using :py:func:`orion.client.report_objective`:
```

Author: Good idea!

.. literalinclude:: ../../../examples/scikitlearn-iris/main.py
   :language: python
   :lines: 32-

In our example, we measure the accuracy of the model to qualify its performance. To get the best
accuracy possible, we need to minimize the difference between 1 and the accuracy to get it as
close to 1 as possible. Otherwise, we'll be minimizing the accuracy which will yield a poor model.

> :py:func:`orion.client.report_results` can be imported using :

Reviewer: ```suggestion
:py:func:`orion.client.report_objective` can be imported using :
```

Author: Good idea!

.. code-block:: python

>    from orion.client import report_results

Reviewer: ```suggestion
   from orion.client import report_objective
```

Updating the script call
^^^^^^^^^^^^^^^^^^^^^^^^
The last missing piece in automating the hyper-parameter optimization of our example model is to
supply Oríon with the values to use for *epsilon*.

We specify the search space in the command line using ``orion~loguniform(1e-5, 1.0)``
as the argument for *espilon*. This argument will tell Oríon to use a log uniform distribution
between ``1e-5`` and ``1`` for the values of *epsilon*.

> Putting everything together, we need to call ./main.py with Oríon. The syntax is the

Reviewer: ```suggestion
Putting everything together, we need to call ``main.py`` with Oríon. The syntax is the
```

Reviewer: Not ``main.py``? :sweat_smile:

Author: I changed it!

Reviewer: But it seams to be still ./main.py

Author: Changed it again...

following: ``$ orion hunt python main.py 'orion~loguniform(1e-5, 1.0)'``. Before executing it on
your terminal, you have to specify the name of the experiment using the ``-n`` option. It is also a
good idea to specify a stopping condition using ``--max-trials`` otherwise the optimization will
not stop unless you interrupt it with :kbd:`ctrl-c`:

.. code-block:: bash

   $ orion hunt -n scitkit-iris-tutorial --max-trials 50 ./main.py 'orion~loguniform(1e-5, 1.0)'

.. warning::
   Make sure you installed the dependencies for the script before running it using ``pip install
   -r requirements.txt``.

Viewing the results
-------------------
Once the optimization reached its stopping condition, you can query Oríon to give you the results
of the optimization with the sub-command ``$ orion info``:

.. code-block:: bash

>    orion info -n scitkit-iris-tutorial

Reviewer: Maybe point to user/monitoring.html#library-api for examples on how to query the results from the DB?

Author: Good idea!

Reviewer: Instead of using the url you can reference it with `` :ref:`library-api-results` ``, the anchor already defined for the section here https://github.com/Epistimio/orion/blame/master/docs/src/user/monitoring.rst#L20

   repository. We invoke this file from index.rst

In this example, we're going to demonstrate how Oríon can be integrated to a minimal model using

> `scikit-learn <https://scikit-learn.org/stable/index.html>`_ on the `iris dataset <https://en.wikipedia.org/wiki/Iris_flower_data_set>`_. The files mentioned in this example are

Reviewer: For iris I would give this [link](http://archive.ics.uci.edu/ml/datasets/Iris) instead so that user are just one click away from downloading it.

> available at https://github.com/Epistimio/orion/tree/master/examples/scikitlearn-iris/.

Reviewer: Maybe `examples/scikitlearn-iris/ <https://github.com/Epistimio/orion/tree/master/examples/scikitlearn-iris/>`_?

Author: That's better.
